lib/view_model/active_record/update_data.rb

```diff
@@ -55,14 +160,158 @@ def initialize(values)
     end
 
     class Replace < self
-      def update_datas
+      def update_datas(referenced_updates)
         values
       end
     end
 
     class Functional < self
-      def update_datas
-        values.flat_map(&:values)
+      def update_datas(referenced_updates)
+        values.reject { |action| action.is_a?(FunctionalUpdate::Remove) }.flat_map(&:values)
+      end
+    end
+
+    class Parser < UpdateParser
+      attr_reader :parse_association
+
+      def initialize(association_data, blame_reference, parse_association:)
+        super(association_data, blame_reference)
+        @parse_association = parse_association
+      end
+
+      def action_schema
+        UpdateData::Schemas::COLLECTION_UPDATE
+      end
+
+      def parse_update_values(values)
+        values.map(&parse_association)
+      end
+
+      def parse_update_action(action)
+        ViewModel::Schemas.verify_schema!(UpdateData::Schemas::UPDATE_ACTION, action)
+        super
+      end
+
+      def parse_remove_action(action)
+        ViewModel::Schemas.verify_schema!(UpdateData::Schemas::REMOVE_ACTION, action)
+        super
+      end
+
+      def parse_append_values(values)
+        values.map(&parse_association)
+      end
+
+      def parse_append_action(action)
+        ViewModel::Schemas.verify_schema!(UpdateData::Schemas::APPEND_ACTION, action)
+        super
+      end
+
+      def parse_replace_values(values)
+        values.map { |child_hash| parse_association.(child_hash) }
+      end
+
+      def new_replace(values)
+        Replace.new(values)
+      end
+
+      def new_functional(values)
+        Functional.new(values)
+      end
+    end
+  end
+
+  class ReferencedCollectionUpdate
+    attr_reader :values
+
+    def initialize(values)
+      @values = values
+    end
+
+    # Values is a list of refs
+    class Replace < self
+      def update_datas(referenced_updates)
+        values.map { |ref| referenced_updates[ref] }
+      end
+    end
+
+    class Functional < self
+      def update_datas(referenced_updates)
+        values
+          .reject { |action| action.is_a?(FunctionalUpdate::Remove) }
+          .flat_map { |val| val.values.map { |ref| referenced_updates[ref] } }
+      end
+    end
+
+    class Parser < UpdateParser
+      attr_reader :valid_reference_keys
+
+      def initialize(association_data, blame_reference, valid_reference_keys:)
+        super(association_data, blame_reference)
+        @valid_reference_keys = valid_reference_keys
+      end
+
+      def action_schema
+        UpdateData::Schemas::REFERENCED_COLLECTION_UPDATE
+      end
+
+      def ref_hash?(value)
```

Comment:
I always instinctively write code as helpers followed by integration, mostly out of habit. This goes directly against using the implementation as private space for help methods following a `private` block. Happy to try a different convention which fits better with the constraints of the language. 

I think the value of naming things is worth keeping this one in the private block.